leaderboard/llvm_codesize/e_greedy/e_greedy.py

```diff
@@ -0,0 +1,115 @@
+    """Determine the best action by trying all possible options and ranking them."""
+
+    def eval_action(fkd: CompilerEnv, action: int) -> ActionReward:
+        """Evaluate the given action."""
+        try:
+            _, reward, _, _ = fkd.step(action)
+        finally:
+            fkd.close()
+        return ActionReward(action=action, reward=reward)
+
+    # Create an ordered list of actions, best first, ordered by descding reward,
+    # using action index as a tie-breaker. Do this by creating n forks of the
+    # environment, one for every action, and evaluting the actions in parallel
+    # threads. Note that calls to fork() occur in the main thread for thread
+    # safety in case of environment restart.
+    futures = (
+        executor.submit(eval_action, env.fork(), action)
+        for action in range(env.action_space.n)
+    )
+    results = (future.result() for future in as_completed(futures))
+    results = sorted(
+        results,
+        key=lambda result: (result.reward, result.action),
+        reverse=True,
+    )
+    return results[0]
+
+
+def e_greedy_search(env: LlvmEnv) -> None:
+    """Run an ϵ-greedy search on an environment."""
+    step_count = 0
+    with ThreadPoolExecutor(max_workers=FLAGS.nproc) as executor:
+        while True:
+            step_count += 1
+
+            if random.random() < FLAGS.epsilon:
+                # Exploratory step. Randomly select and apply an action.
+                action = env.action_space.sample()
+                _, reward, done, _ = env.step(action)
+                logging.debug(
+                    "Step %d, exploratory action %s, reward %.4f, cumulative %.4f",
+                    step_count,
+                    env.action_space.flags[action],
+                    reward,
+                    env.episode_reward,
+                )
+            else:
+                # Select the best reward and apply it, or terminate the search
+                # if no positive reward is attainable.
+                best = select_best_action(env, executor)
+                if best.reward <= 0:
+                    logging.debug(
+                        "Greedy search terminated after %d steps, "
+                        "no further reward attainable",
+                        step_count,
+                    )
+                    done = True
+                else:
+                    _, reward, done, _ = env.step(best.action)
+                    logging.debug(
+                        "Step %d, greedy action %s, reward %.4f, cumulative %.4f",
+                        step_count,
+                        env.action_space.flags[best.action],
+                        reward,
+                        env.episode_reward,
+                    )
+                    if env.reward_space.deterministic and reward != best.reward:
+                        logging.warning(
+                            "Action %s produced different reward on replay, %.4f != %.4f",
```

Comment:
How is this possible?

Reply:
This is possible if one of the underlying actions is nondeterministic. This has been found in the past, e.g.
https://github.com/facebookresearch/CompilerGym/issues/46